### ISSUE TITLE
XLA Runtime error using axis_index inside shard_map

### ISSUE THREAD

**Issue by jlottes** [google/jax#15780]
### Description

Possibly similar to #15723, but a different error. Running the following:
```python
import functools, os
os.environ['XLA_FLAGS'] = '--xla_force_host_platform_device_count=2'
import jax
from jax.experimental import mesh_utils
from jax.experimental.shard_map import shard_map

devices = mesh_utils.create_device_mesh((jax.device_count(),))
mesh = jax.sharding.Mesh(devices, axis_names=('i'))
spec = jax.sharding.PartitionSpec('i')

@jax.jit
def f(x, cond):
  @functools.partial(shard_map, mesh=mesh, in_specs=spec, out_specs=spec, check_rep=False)
  def g(x):
    idx = jax.lax.axis_index(axis_name='i')
    return jax.lax.cond(cond,
                        lambda: x < idx,
                        lambda: x < 5)
  return g(x)

f(jax.numpy.arange(2), True)
```
results in the error
```
XlaRuntimeError: UNIMPLEMENTED: PartitionId instruction is not supported for SPMD partitioning since the meaning is ambiguous -- whether the instruction is replicated or the data is replicated, and if the latter which data is replicated.
```
The error goes away if `idx` is inlined, or the `lax.cond` is replaced with either branch. My suspicion is that the error arises because `idx` is defined but not used in the second branch. Not sure if this is an XLA issue, or an issue with JAX lowering.

```python
print(f.lower(jax.numpy.arange(2), True).as_text())
```
shows the HLO
```
module @jit_f attributes {mhlo.num_partitions = 2 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<2xi32> {jax.arg_info = "x", mhlo.sharding = "{replicated}"}, %arg1: tensor<i1> {jax.arg_info = "cond", mhlo.sharding = "{replicated}"}) -> (tensor<2xi1> {jax.result_info = ""}) {
    %0 = stablehlo.custom_call @Sharding(%arg1) {mhlo.sharding = "{replicated}"} : (tensor<i1>) -> tensor<i1>
    %1 = stablehlo.custom_call @SPMDFullToShardShape(%0) {mhlo.sharding = "{manual}"} : (tensor<i1>) -> tensor<i1>
    %2 = stablehlo.custom_call @Sharding(%arg0) {mhlo.sharding = "{devices=[2]0,1}"} : (tensor<2xi32>) -> tensor<2xi32>
    %3 = stablehlo.custom_call @SPMDFullToShardShape(%2) {mhlo.sharding = "{manual}"} : (tensor<2xi32>) -> tensor<1xi32>
    %4 = stablehlo.constant dense<1> : tensor<ui32>
    %5 = stablehlo.constant dense<2> : tensor<ui32>
    %6 = stablehlo.partition_id : tensor<ui32>
    %7 = stablehlo.divide %6, %4 : tensor<ui32>
    %8 = stablehlo.remainder %7, %5 : tensor<ui32>
    %9 = stablehlo.convert %8 : (tensor<ui32>) -> tensor<i32>
    %10 = stablehlo.convert %1 : (tensor<i1>) -> tensor<i32>
    %11 = "stablehlo.case"(%10) ({
      %14 = stablehlo.constant dense<5> : tensor<i32>
      %15 = stablehlo.broadcast_in_dim %14, dims = [] : (tensor<i32>) -> tensor<1xi32>
      %16 = stablehlo.compare  LT, %3, %15,  SIGNED : (tensor<1xi32>, tensor<1xi32>) -> tensor<1xi1>
      stablehlo.return %16 : tensor<1xi1>
    }, {
      %14 = stablehlo.broadcast_in_dim %9, dims = [] : (tensor<i32>) -> tensor<1xi32>
      %15 = stablehlo.compare  LT, %3, %14,  SIGNED : (tensor<1xi32>, tensor<1xi32>) -> tensor<1xi1>
      stablehlo.return %15 : tensor<1xi1>
    }) : (tensor<i32>) -> tensor<1xi1>
    %12 = stablehlo.custom_call @Sharding(%11) {mhlo.sharding = "{manual}"} : (tensor<1xi1>) -> tensor<1xi1>
    %13 = stablehlo.custom_call @SPMDShardToFullShape(%12) {mhlo.sharding = "{devices=[2]0,1}"} : (tensor<1xi1>) -> tensor<2xi1>
    return %13 : tensor<2xi1>
  }
}
```

### Which accelerator(s) are you using?

CPU

**Comment by rajasekharporeddy** [comment]
Hi @jlottes

It appears the issue has been resolved in JAX version 0.4.31. I tested the mentioned code on Colab and it executed without any error.

```python
import functools, os
os.environ['XLA_FLAGS'] = '--xla_force_host_platform_device_count=2'
import jax
from jax.experimental import mesh_utils
from jax.experimental.shard_map import shard_map

devices = mesh_utils.create_device_mesh((jax.device_count(),))
mesh = jax.sharding.Mesh(devices, axis_names=('i'))
spec = jax.sharding.PartitionSpec('i')

@jax.jit
def f(x, cond):
  @functools.partial(shard_map, mesh=mesh, in_specs=spec, out_specs=spec, check_rep=False)
  def g(x):
    idx = jax.lax.axis_index(axis_name='i')
    return jax.lax.cond(cond,
                        lambda: x < idx,
                        lambda: x < 5)
  return g(x)

f(jax.numpy.arange(2), True)
```

Output:

```
Array([False, False], dtype=bool)
```

Please find the attached [gist](https://colab.sandbox.google.com/gist/rajasekharporeddy/4f38b94f08968cb3b16713573383209a/-15780.ipynb) for reference.

Thank you.

**Comment by mattjj** [comment]
Thank you, @rajasekharporeddy ! It makes sense to me that this was fixed in the interim, yet we neglected to update this bug.
